### ISSUE TITLE
Using Pushy with BouncyCastleProvider

### ISSUE THREAD

**Issue by dorheini** [relayrides/pushy#247]
Hi.
I'm using BouncyCastle as security provider and the following code throws KeyStoreException when instantiating ApnsClient with a *.p12 certificate file:
`if (keyStore.size() != 1) {
            throw new KeyStoreException("Key store must contain exactly one entry, and that entry must be a private key entry.");
        } `

The keyStore.size() value is  2.

With sun.security.provider.Sun as security provider the above code works but we use BouncyCastleProvider as security provider.

**Comment by jchambers** [comment]
@dorheini Could you please let us know which version of BouncyCastle you're using and by which method you've installed it as your security provider? I doubt the issue is version- or method-specific, but it often helps to be specific when trying to recreate this kind of issue.

Thanks!

**Comment by dorheini** [comment]
@jchambers The BouncyCastle version is 1.47 (but also tested with 1.53 & 1.54).
I installed BouncyCastle as security provider by changing the $JAVA_HOME/jre/lib/security/java.security file (setting security.provider.1=org.bouncycastle.jce.provider.BouncyCastleProvider) and adding the bcprov-jdk15on-1.47.jar (also 1.53 and 1.54 as mentioned above) to $JAVA_HOME/jre/lib/ext

**Comment by jchambers** [comment]
@dorheini I'm having a hard time reproducing this issue; I've set BouncyCastle 1.54 as my primary provider, but all of the tests still run and pass (and see a single-entry keystore) without any problems. Did the tests fail for you, or was it only an issue outside of the context of tests?

**Comment by jchambers** [comment]
I put together a small app that should dump what your security provider thinks is in your key store. Could you please share the output of this app with and without BouncyCastle as your provider

```java
import java.io.File;
import java.io.FileInputStream;
import java.security.KeyStore;
import java.security.Provider;
import java.security.Security;
import java.util.Enumeration;

public class DumpP12App {

    private static final File P12_FILE = new File("/path/to/your/file.p12");
    private static final char[] KEYSTORE_PASSWORD = "your-password-here".toCharArray();

    public static void main(final String[] args) throws Exception {
        final KeyStore p12KeyStore = KeyStore.getInstance("PKCS12");
        p12KeyStore.load(new FileInputStream(P12_FILE), KEYSTORE_PASSWORD);

        System.out.println("Security providers:");

        for (final Provider provider : Security.getProviders()) {
            System.out.format("\t- %s\n", provider.getName());
        }

        System.out.println();

        final KeyStore.PasswordProtection passwordProtection = new KeyStore.PasswordProtection(KEYSTORE_PASSWORD);
        final Enumeration<String> aliases = p12KeyStore.aliases();

        System.out.format("Key store contents (%d entries):\n", p12KeyStore.size());

        while (aliases.hasMoreElements()) {
            final String alias = aliases.nextElement();
            System.out.format("\t-%s: %s\n", alias, p12KeyStore.getEntry(alias, passwordProtection).getClass().getName());
        }
    }
}
```

Thanks!

**Comment by jchambers** [comment]
Never mind! I figured it out. BouncyCastle is seeing a trusted certificate entry that the Sun provider isn't:

```
Key store contents (2 entries):
	- Apple Push Services: [redacted]: java.security.KeyStore$TrustedCertificateEntry
	- [redacted]: java.security.KeyStore$PrivateKeyEntry
```

It appears this does *not* happen with the P12 files we use for unit tests, so I'm sorry I wasn't able to reproduce the problem sooner. I do think this is a "bug" in that we're being a bit more aggressive than we need to be. I'm going to relax the check so that it's okay to have multiple entries in the key store, but only one of those entries can be a `PrivateKeyEntry`.

**Comment by jchambers** [comment]
This is fixed in 8d7acaa. Please let us know if you have any further trouble!
